RecoEgamma/EgammaTools/src/HGCalShowerShapeHelper.cc

```diff
@@ -0,0 +1,310 @@
+
+    if (hitLayer > nLayer_) {
+      continue;
+    }
+
+    if (hitId.det() != subDet_) {
+      continue;
+    }
+
+    if (pfRecHitPtrMap_.find(hitId.rawId()) == pfRecHitPtrMap_.end()) {
+      continue;
+    }
+
+    reco::PFRecHit recHit = *pfRecHitPtrMap_[hitId.rawId()];
+
+    if (recHit.energy() < minHitE_) {
+      continue;
+    }
+
+    if (recHit.pt2() < minHitET2_) {
+      continue;
+    }
+
+    // Fill the vectors
+    hitsAndFracs_.push_back(hnf);
+    hitEnergies_.push_back(recHit.energy());
+    hitEnergiesWithFracs_.push_back(recHit.energy() * hitEfrac);
+  }
+}
+
+void HGCalShowerShapeHelper::setLayerWiseStuff() {
+  layerEnergies_.clear();
+  layerEnergies_.resize(nLayer_);
+
+  layerCentroids_.clear();
+  layerCentroids_.resize(nLayer_);
+
+  centroid_.SetXYZ(0, 0, 0);
+
+  int iHit = -1;
+  double totalW = 0;
+
+  // Compute the centroid per layer
+  for (const auto &hnf : hitsAndFracs_) {
+    iHit++;
+
+    DetId hitId = hnf.first;
+
+    double weight = hitEnergies_[iHit];
+    totalW += weight;
+
+    const auto &hitPos = recHitTools_.getPosition(hitId);
+    ROOT::Math::XYZVector hitXYZ(hitPos.x(), hitPos.y(), hitPos.z());
+
+    centroid_ += weight * hitXYZ;
+
+    int hitLayer = recHitTools_.getLayer(hitId) - 1;
+
+    layerEnergies_[hitLayer] += weight;
+    layerCentroids_[hitLayer] += weight * hitXYZ;
+  }
+
+  int iLayer = -1;
+
+  for (auto &centroid : layerCentroids_) {
+    iLayer++;
+
+    if (layerEnergies_[iLayer]) {
+      centroid /= layerEnergies_[iLayer];
+    }
+  }
+
+  if (totalW) {
+    centroid_ /= totalW;
+  }
+}
+
+double HGCalShowerShapeHelper::getCellSize(DetId detId) {
+  double siThickness = recHitTools_.getSiThickness(detId);
+
+  // HD wafers
+  if (siThickness < 150) {
+    return 0.465;
+  }
+
+  // LD wafers
+  else {
+    return 0.698;
+  }
+}
+
+double HGCalShowerShapeHelper::getRvar(double cylinderR, double energyNorm, bool useFractions, bool useCellSize) {
+  if (hitsAndFracs_.empty()) {
+    return 0;
+  }
+
+  assert(energyNorm > 0);
+
+  double cylinderR2 = cylinderR * cylinderR;
+
+  double Rvar = 0;
+
+  std::vector<double>::iterator hitEnergyIter;
+
+  if (useFractions) {
+    hitEnergyIter = hitEnergiesWithFracs_.begin();
+  }
+
+  else {
+    hitEnergyIter = hitEnergies_.begin();
+  }
+
+  hitEnergyIter--;
+
+  for (const auto &hnf : hitsAndFracs_) {
+    hitEnergyIter++;
+
+    DetId hitId = hnf.first;
+
+    int hitLayer = recHitTools_.getLayer(hitId) - 1;
+
+    const auto &hitPos = recHitTools_.getPosition(hitId);
+    ROOT::Math::XYZVector hitXYZ(hitPos.x(), hitPos.y(), hitPos.z());
+
+    auto distXYZ = hitXYZ - layerCentroids_[hitLayer];
+
+    double r2 = distXYZ.x() * distXYZ.x() + distXYZ.y() * distXYZ.y();
+
+    // Including the cell size seems to make the variable less sensitive to the HD/LD transition region
+    if (useCellSize) {
+      if (sqrt(r2) > cylinderR + getCellSize(hitId)) {
+        continue;
+      }
+    }
+
```

Comment:
was this code formatting (I really hope not as I find the if /else are disconected this way and harder to follow)

Reply:
Have changed it to `else if (...) {`